feat(kms): change default key policy to align with KMS best practices (under feature flag) (#11918)
aws/aws-cdk: packages/@aws-cdk/aws-kms/README.md

no-ops, and `addAlias` and `aliasTargetKey` will fail, as the imported alias does not
have a reference to the underlying KMS Key.

## Key Policies

Controlling access and usage of KMS Keys requires the use of key policies (resource-based policies attached to the key);
this is in contrast to most other AWS resources where access can be entirely controlled with IAM policies,
and optionally complemented with resource policies. For more in-depth understanding of KMS key access and policies, see

* https://docs.aws.amazon.com/kms/latest/developerguide/control-access-overview.html
* https://docs.aws.amazon.com/kms/latest/developerguide/key-policies.html

KMS keys can be created to trust IAM policies. This is the default behavior for both the KMS APIs and in
the console. This behavior is enabled by the '@aws-cdk/aws-kms:defaultKeyPolicies' feature flag,

> which is set for all new projects; for existing projects, this same behavior can be enabled by:

Reviewer: ```suggestion
which is set for all new projects; for existing projects, this same behavior can be enabled by passing the `trustAccountIdentities` property as `true` when creating the key:
```

```ts
new kms.Key(stack, 'MyKey', { trustAccountIdentities: true });
```

With either the `defaultKeyPolicies` feature flag set, or `trustAccountIdentities` set,
the Key will be given the following default key policy:

```json
{
  "Effect": "Allow",
  "Principal": {"AWS": "arn:aws:iam::111122223333:root"},
  "Action": "kms:*",
  "Resource": "*"
}
```

This policy grants full access to the key to the root account user.
This enables the root account user -- via IAM policies -- to grant access to other IAM principals.
With the above default policy, future permissions can be added to either the key policy or IAM principal policy.

```ts
const key = new kms.Key(stack, 'MyKey');

> const user = new iam.user(stack, 'MyUser');

Reviewer: ```suggestion
const user = new iam.User(stack, 'MyUser');
```

key.grantEncrypt(user); // Adds encrypt permissions to user policy; key policy is unmodified.
```

Adopting the default KMS key policy (and so trusting account identities)
solves many issues around cyclic dependencies between stacks.
Without this default key policy, future permissions must be added to both the key policy and IAM principal policy,
which can cause cyclic dependencies if the permissions cross stack boundaries.
(For example, an encrypted bucket in one stack, and Lambda function that accesses it in another.)

### Appending to or replacing the default key policy

The default key policy can be amended or replaced entirely, depending on your use case and requirements.
A common addition to the key policy would be to add other key admins that are allowed to administer the key
(e.g., change permissions, revoke, delete). Additional key admins can be specified at key creation or after
via the `addAdmin` method.

```ts
const myTrustedAdminRole = iam.Role.fromRoleArn(stack, 'TrustedRole', 'arn:aws:iam:....');
const key = new kms.Key(stack, 'MyKey', {
  admins: [myTrustedAdminRole],
});

const secondKey = new kms.Key(stack, 'MyKey2');

> secondKey.addAdmin(myTrustedAdminRole);

Reviewer: I _think_ this should be `grantAdmin()`, not `addAdmin()`, according to the code below...?

```

Alternatively, a custom key policy can be specified, which will replace the default key policy.

> **Note**: In applications without the '@aws-cdk/aws-kms:defaultKeyPolicies' feature flag set,

> or `trustedAccountIdentities` set to false, specifying a policy at key creation _appends_ the

Reviewer: I think the wording here is slightly unclear... how about:

```suggestion
and with `trustedAccountIdentities` set to `false`, which is the default, specifying a policy at key creation _appends_ the
```
